### ISSUE TITLE
run jobs by monitoring consul kv

### ISSUE THREAD

**Issue by ketzacoatl** [hashicorp/nomad#1038]
I have a feature request, one that I believe will prove very interesting and useful as Nomad (and Consul) mature. If Nomad were to accept a config param that was the KV path (prefix) in Consul to monitor, it could watch those keys for changes, and if Nomad considered each key a job spec, Nomad could run jobs straight from Consul.

In other words, the behavior would be like:
- when the nomad servers have established a quorum and have a leader, the leader could be responsible (or make another server responsible) for monitoring the special KV path.
- for any subkeys in that path, nomad would attempt to run the job using the contents of the subkey as the job spec
- if or when there were any changes in the subkeys at that path, nomad would update the running jobs
- if a subkey was present, but then goes away, nomad would stop the corresponding job that is running

I ended up realizing this as an elegant solution to a peripheral problem I have been thinking on for a while. The general gist is that this pairing of Consul and Nomad provides for a very elegant bootstrap for a highly automated cluster of services. For example, there is a bit of a nasty chicken/egg scenario that arises when you try to automate managing jobs with Nomad - Nomad is the one responsible for keeping N instances of M services online, but if we want to automate running some of these jobs (say like a group of management tasks), we need something to tell Nomad what to do. But I need to ensure that new service stays online.. so rather than tell Nomad what what to do, I'd like Nomad to have the ability to look for some work when it comes online. This also helps me avoid race conditions in ensuring these tasks are  online.

Let's say I start a cluster of three nodes and bootstrap them as a cluster of Consul Leaders, and Nomad Servers. Once they have achieved their quorum, Nomad is ready for jobs. Let's say we then want to ensure Nomad runs N services which are responsible for managing the clusters (of servers and workers). We could distribute the job spec to those Nomad servers via a node init in some way, and `nomad run` that during provisioning, but all ways I have considered are either not elegant or just plain wrong for some given use case (usually a source of race conditions). If Nomad were to look to Consul and run the jobs it sees in the KV, we would have a clean and very flexible solution for any situation where you would like to streamline and push the limits within the automated hosting platform.

EDIT: While I do not have the golang chops to pull this off, I would be more than happy to run tests and help out in any other way that I can.

**Comment by sean-** [comment]
This is an interesting consideration that dovetails with hashicorp/consul-template#570.  Thank you for taking the time to submit this issue.

**Comment by dadgar** [comment]
Hey this is somewhat out of scope. It can be easily handled with consul-template by rendering the contents to a file and then running a script to submit them to Nomad.

**Comment by skyrocknroll** [comment]
We are doing that in our prod and qe.
